### PR TITLE
Professor/ch3819/external ipv4 addresses should sort before

### PR DESCRIPTION
**Pull request by macpie**

### CODE REVIEW

File: src/libp2p_transport.erl [hunk 1]
```diff
@@ -41,21 +41,29 @@ for_addr(TID, Addr) ->
...
+        libp2p_config:lookup_transports(TID)
     ),
-
-    case lists:sort(Matches) of
+    case Matches of
         [] ->
             {error, {unsupported_address, Addr}};
-        [{_, Matched, {Transport, Pid}}|_] ->
-            {ok, Matched, {Transport, Pid}}
+        [{Matched, {Transport, Pid}}] ->
```
Comment (macpie) [inline src/libp2p_transport.erl]: We should only match one transport

**Comment by macpie** [comment]
Each transport can now, prioritize each individual address. This allow tcp to make the difference for `rfc1918` IPs.

File: src/proxy/libp2p_transport_proxy.erl [hunk 2]
```diff
@@ -28,12 +27,9 @@ start_listener(_Pid, _Addr) ->
...
     false.
 
--spec sort_addrs([string()]) -> [string()].
+-spec sort_addrs([string()]) -> [{integer(), string()}].
 sort_addrs(Addrs) ->
-    Addrs.
-
--spec priority() -> integer().
-priority() -> 99.
+    [{5, A} || A <- Addrs].
```
Comment (Vagabond) [inline src/proxy/libp2p_transport_proxy.erl]: Why not keep this as 99?

File: src/libp2p_transport.erl [hunk 3]
```diff
@@ -41,21 +41,29 @@ for_addr(TID, Addr) ->
...
+            {ok, Matched, {Transport, Pid}};
+        [{_Matched, {_Transport, _Pid}}|_] ->
+            {error, {multiple_match_address, Addr}}
     end.
 
+%%--------------------------------------------------------------------
+%% @doc
+%% Addresses are sorted by priority randing from 1 to 5 (1 highest and 5 lowest priority)
```
Comment (Vagabond) [inline src/libp2p_transport.erl]: should be `ranking`

File: src/proxy/libp2p_transport_proxy.erl (hunk 2 above)
Comment (macpie) [reply src/proxy/libp2p_transport_proxy.erl]: idk just wanted a smaller range  99 seemed just useless

Comment (Vagabond) [reply src/proxy/libp2p_transport_proxy.erl]: I guess this will never pass `match_addr` so it's academic

Comment (macpie) [reply src/proxy/libp2p_transport_proxy.erl]: yes

**Comment by Vagabond** [comment]
This looks good, let's give it a try in the field.

**Pull request merged**